src/modules/airdrop/services/airdrop-service.ts

```diff
@@ -123,6 +132,89 @@ export class AirdropService {
     };
   }
 
+  async processMerkleDistributorRecipientsFile(file?: Express.Multer.File) {
+    if (!file) return;
+
+    const recipientsFile = await readFile(file.path, { encoding: "utf8" });
+    const recipientsJson = JSON.parse(recipientsFile);
+
+    try {
+      return await this.dataSource.transaction(async (entityManager) => {
+        const window = await entityManager
+          .createQueryBuilder()
+          .insert()
+          .into(MerkleDistributorWindow)
+          .values({
+            chainId: recipientsJson["chainId"],
+            rewardToken: recipientsJson["rewardToken"],
+            windowIndex: recipientsJson["windowIndex"],
+            rewardsToDeposit: recipientsJson["rewardsToDeposit"],
+            merkleRoot: recipientsJson["merkleRoot"],
+          })
+          .execute();
+        const windowId = window.identifiers[0].id;
+
+        for (const recipient of Object.values(recipientsJson["recipientsWithProofs"])) {
+          await entityManager
+            .createQueryBuilder()
+            .insert()
+            .into(MerkleDistributorRecipient)
+            .values({
+              merkleDistributorWindowId: windowId,
+              address: recipient["account"],
+              amount: recipient["amount"],
+              accountIndex: recipient["accountIndex"],
+              proof: recipient["proof"],
+              payload: recipient["metadata"],
+            })
+            .execute();
+        }
```

Comment:
AFAICT the current behavior for an edge case where an insert fails for a single recipient would require us to delete the entries manually before re-uploading the fixed file right?

Reply:
Could be nice to make this atomic somehow but maybe also over-engineered

Reply:
@dohaki 
It is an atomic operation 👍 These insertions are done using a [SQL transaction](https://www.postgresql.org/docs/current/tutorial-transactions.html) (see [line 142](https://github.com/across-protocol/scraper-api/blob/d3002023cf15c479e87070430448cfa800ea984c/src/modules/airdrop/services/airdrop-service.ts#L142) where the transaction is started). If an error is thrown inside this big callback, then all the inserts are reverted. 

PS. This is the easiest way to run SQL transactions using TypeORM and basically the library handles all the operations: start, commit and revert. TypeORM can also let you handle transactions manually, [see example here](https://github.com/typeorm/typeorm/blob/master/docs/transactions.md#using-queryrunner-to-create-and-control-state-of-single-database-connection)

Reply:
awesome thanks for the clarification